packages/lib-core/src/extensions/actions.ts

```diff
@@ -0,0 +1,119 @@
+import type { ReactNode } from 'react';
+import type { ExtensionHook, ExtensionK8sKindVersionModel } from '../api/common-types';
+import type { CodeRef, Extension } from '../types/extension';
+import type { AccessReviewResourceAttributes } from './core-types';
+
+/** ActionProvider contributes a hook that returns list of actions for specific context */
+export type ActionProvider = Extension<
+  'core.action/provider',
+  {
+    /** The context ID helps to narrow the scope of contributed actions to a particular area of the application. Ex - topology, helm */
+    contextId: string | 'resource';
+    /** A react hook which returns actions for the given scope.
+     * If contextId = `resource` then the scope will always be a K8s resource object
+     * */
+    provider: CodeRef<ExtensionHook<Action[]>>;
+  }
+>;
+
+/** ResourceActionProvider contributes a hook that returns list of actions for specific resource model */
+export type ResourceActionProvider = Extension<
+  'core.action/resource-provider',
+  {
+    /** The model for which this provider provides actions for. */
+    model: ExtensionK8sKindVersionModel;
+    /** A react hook which returns actions for the given resource model */
+    provider: CodeRef<ExtensionHook<Action[]>>;
+  }
+>;
+
+/** ActionGroup contributes an action group that can also be a submenu */
+export type ActionGroup = Extension<'core.action/group', Group>;
+
+/** ActionFilter can be used to filter an action */
+export type ActionFilter = Extension<
+  'core.action/filter',
+  {
+    /** The context ID helps to narrow the scope of contributed actions to a particular area of the application. Ex - topology, helm */
+    contextId: string | 'resource';
+    /** A function which will filter actions based on some conditions.
+     * scope: The scope in which actions should be provided for.
+     * Note: hook may be required if we want to remove the ModifyCount action from a deployment with HPA
+     * */
+    filter: CodeRef<(scope: unknown, action: Action) => boolean>;
```

Comment:
I guess `scope` was originally typed as `any` and replaced with `unknown` type.

Ideally we would provide a better type here, it should be more apparent from how the extension is used.

Reply:
I'm not sure I can provide much information what the `scope` can be here :disappointed:

Reply:
The `scope` is different for every context. For example in topology the `scope` would be a scene graph element, while for a resource page, the scope would be the resource object itself.
The intent was for reusing the same extension for supplying actions in different contexts. But maybe it makes more sense to use completely different extension types that inherit the same structure ?

Reply:
`scope` should follow the same typing as `ExtensionHook` which accepts a typed parameter in the function call.
Unsure if `unknown` or `any` is more appropriate here but the type can be cast to a specific type based on the `contextId`.

Reply:
resource actions will be important to bring over for supporting details and list pages which present resource actions to the user.

Reply:
So the function type can be written as:
```TS
type FilterType<TType> = (scope: TType, action: Action) => boolean
```

Reply:
Not sure the type necessarily helps in this case. The user is a plugin. They define their functionality -- the caller is the App (in this case) and they will provide what they provide based on where they call it.

The Generic type won't translate well between the two I don't think. It's a weird concept. Not sure why we have such a loose extension. I think we can leave it as `unknown` for now imo. The plugin can type it when they know the use-case they built it for for their own internal function typing.

These action types are weird 😕

Reply:
> The Generic type won't translate well between the two I don't think.

Right, if it's generic, both app and plugin need to make an assumption about it.

Reply:
I feel it's still better typed with a generic type param (falling back to `unknown`) but we can add it later.